x-pack/libbeat/management/managerV2.go

```diff
@@ -0,0 +1,300 @@
+	log := logp.NewLogger(lbmanagement.DebugK)
+	m := &BeatV2Manager{
+		config:   config,
+		logger:   log.Named("V2-manager"),
+		registry: registry,
+		units:    make(map[string]*client.Unit),
+	}
+
+	if config.Enabled {
+		m.client = agentClient
+	}
+	return m, nil
+}
+
+// ================================
+// Beats central management interface implementation
+// ================================
+
+// UpdateStatus updates the manager with the current status for the beat.
+func (cm *BeatV2Manager) UpdateStatus(status lbmanagement.Status, msg string) {
+	updateState := client.UnitState(status)
+	_ = cm.getUnit(cm.mainUnit).UpdateState(updateState, msg, cm.payload)
+}
+
+// Enabled returns true if config management is enabled.
+func (cm *BeatV2Manager) Enabled() bool {
+	return cm.config.Enabled
+}
+
+// SetStopCallback sets the callback to run when the manager want to shutdown the beats gracefully.
+func (cm *BeatV2Manager) SetStopCallback(stopFunc func()) {
+	cm.stopFunc = stopFunc
+}
+
+// Start the config manager.
+func (cm *BeatV2Manager) Start() error {
+	if !cm.Enabled() {
+		return fmt.Errorf("V2 Manager is disabled")
+	}
+	err := cm.client.Start(context.Background())
+	if err != nil {
+		return fmt.Errorf("error starting connection to client")
+	}
+
+	go cm.unitListen()
+	cm.isRunning = true
+	return nil
+}
+
+// Stop stops the current Manager and close the connection to Elastic Agent.
+func (cm *BeatV2Manager) Stop() {
+	cm.unitsMut.Lock()
+	defer cm.unitsMut.Unlock()
+	main, ok := cm.units[cm.mainUnit]
+	if ok {
+		cm.stopBeat(main)
+	}
+}
+
+// CheckRawConfig is currently not implemented for V1.
+func (cm *BeatV2Manager) CheckRawConfig(cfg *conf.C) error {
+	// This does not do anything on V1 or V2, but here we are
+	return nil
+}
+
+func (cm *BeatV2Manager) RegisterAction(action client.Action) {
+	cm.unitsMut.Lock()
+	defer cm.unitsMut.Unlock()
+	cm.units[cm.mainUnit].RegisterAction(action)
+}
+
+func (cm *BeatV2Manager) UnregisterAction(action client.Action) {
+	cm.unitsMut.Lock()
+	defer cm.unitsMut.Unlock()
+	cm.units[cm.mainUnit].UnregisterAction(action)
+}
+
+func (cm *BeatV2Manager) SetPayload(payload map[string]interface{}) {
+	cm.payload = payload
+}
+
+// ================================
+// Unit manager
+// ================================
+
+func (cm *BeatV2Manager) addUnit(unit *client.Unit) {
+	cm.unitsMut.Lock()
+	cm.units[unit.ID()] = unit
+	cm.unitsMut.Unlock()
+}
+
+func (cm *BeatV2Manager) getUnit(ID string) *client.Unit {
+	cm.unitsMut.Lock()
+	defer cm.unitsMut.Unlock()
+	return cm.units[ID]
+
+}
+
+func (cm *BeatV2Manager) deleteUnit(unit *client.Unit) {
+	cm.unitsMut.Lock()
+	delete(cm.units, unit.ID())
+	cm.unitsMut.Unlock()
+}
+
+// ================================
+// Private V2 implementation
+// ================================
+
+func (cm *BeatV2Manager) unitListen() {
+	cm.logger.Debugf("Listening for agent unit changes")
+	for {
+		change := <-cm.client.UnitChanges()
```

Comment:
Do you need to detect that the channel was closed here? What does this do if the agent crashes unexpectedly?

Reply:
That's what I was wondering. The pre-existing reloader doesn't provide any kind of channel, but I think the solution is probably to just have the `Stop()` function from the reloader interface send a done signal.

Reply:
This channel will not be closed even if the client is stopped. The channel is created at initialization and always stays the same. You can stop and stop the client and it would be the same channel.

If you need to close the goroutine that is listening for changes you should pass in a context and select on the Done() channel to close the goroutine.